include/oneapi/dpl/pstl/execution_impl.h

```diff
@@ -111,42 +79,57 @@ template <typename _ExecutionPolicy>
+// backend selector with tags
+//------------------------------------------------------------------------
+
+template <class _Policy, class... _IteratorTypes>
+struct __vectorable_tag
 {
-    return __internal::__lazy_and(__exec.__allow_vector(),
-                                  typename __internal::__is_random_access_iterator<_IteratorTypes...>::type());
-}
+    using __is_vector = __conjunction<__allow_unsequenced<_Policy>,
+                                      typename __internal::__is_random_access_iterator<_IteratorTypes...>>;
+};
 
-template <typename _ExecutionPolicy, typename... _IteratorTypes>
-auto
-__is_parallelization_preferred(_ExecutionPolicy& __exec)
-    -> decltype(__internal::__lazy_and(__exec.__allow_parallel(),
-                                       typename __internal::__is_random_access_iterator<_IteratorTypes...>::type()))
+template <class _Policy, class... _IteratorTypes>
+struct __serial_tag : __vectorable_tag<_Policy, _IteratorTypes...>
 {
-    return __internal::__lazy_and(__exec.__allow_parallel(),
-                                  typename __internal::__is_random_access_iterator<_IteratorTypes...>::type());
-}
+};
 
-template <typename policy, typename... _IteratorTypes>
-struct __prefer_unsequenced_tag
+template <class _Policy, class... _IteratorTypes>
+struct __parallel_tag : __vectorable_tag<_Policy, _IteratorTypes...>
 {
-    static constexpr bool value = __internal::__allow_unsequenced<policy>::value &&
-                                  __internal::__is_random_access_iterator<_IteratorTypes...>::value;
-    typedef ::std::integral_constant<bool, value> type;
 };
 
-template <typename policy, typename... _IteratorTypes>
-struct __prefer_parallel_tag
+template <class _Policy, class... _IteratorTypes>
+struct __parallel_forward_tag
 {
-    static constexpr bool value = __internal::__allow_parallel<policy>::value &&
-                                  __internal::__is_random_access_iterator<_IteratorTypes...>::value;
-    typedef ::std::integral_constant<bool, value> type;
+    using __is_vector = ::std::false_type;
 };
 
+template <typename _Policy, typename... _IteratorTypes>
+using __tag_type =
+    typename ::std::conditional<__internal::__is_random_access_iterator<_IteratorTypes...>::value,
+                                __parallel_tag<_Policy, _IteratorTypes...>,
+                                typename ::std::conditional<__is_forward_iterator<_IteratorTypes...>::value,
+                                                            __parallel_forward_tag<_Policy, _IteratorTypes...>,
+                                                            __serial_tag<_Policy, _IteratorTypes...>>::type>::type;
+
+template <typename _Policy, class... _IteratorTypes>
+typename ::std::enable_if<!__allow_parallel<_Policy>::value,
+                          __serial_tag<_Policy, typename ::std::decay<_IteratorTypes>::type...>>::type
+__select_backend(_Policy&&, _IteratorTypes&&...)
```

Comment:
Should we add a macro to turn these selectors off, so that another backend can customize the behaviour for the standard execution policies and iterators without changes in this file?

Reply:
I believe that it's on the side of the customization implementers. They add such a macro and substitute our `select_backend` with their implementation. Anyway, the approach would be ratified during the negotiation with our stakeholders.